### PR TITLE
[collectd 6] write_prometheus plugin: Fix handling of multiple resources.

### PR DESCRIPTION
**Pull request by octo**
This fixes the problem with multiple resources outlined in #4283.

* Adds a new "end to end" test that writes metric families via `prom_write()` and verifies the output of `format_text()`.
* Prints the `job` and `instance` resource attributes for metrics without labels. Previously such metrics incorrectly triggered an early return and where thus printed without these important attributes.
* Introduces `prometheus_metric_family_t` and `prometheus_metric_t`, which structure data in the way required by the formatting. This fixes the original problem and significantly simplifies memory management.

ChangeLog: Write Prometheus plugin: Bugs in the handling of metrics from multiple resources as well as printing of metrics without labels have been fixed.

### CODE REVIEW

File: src/write_prometheus.c
Comment (eero-t) [inline src/write_prometheus.c:450]: IMHO it would be much nicer for `write_prometheus_test.c`  to include `write_prometheus.c` (with `Makefile.am` updated accordingly), instead of changing all these functions to non-static.

File: src/write_prometheus_test.c
Comment (eero-t) [inline src/write_prometheus_test.c:445]: Unnecessary if test `#include`s the tested plugin code.

**Review by eero-t — CHANGES_REQUESTED** [review]

File: src/write_prometheus.c
Comment (octo) [reply src/write_prometheus.c:471]: ✅ done

File: src/write_prometheus_test.c
Comment (octo) [reply src/write_prometheus_test.c]: ✅ done

File: src/write_prometheus_test.c
Comment (eero-t) [inline src/write_prometheus_test.c:465]: On same line, as in next array item:
```suggestion
	    "# collectd/write_prometheus " PACKAGE_VERSION " at example.com\n",
```

**Review by eero-t — CHANGES_REQUESTED** [review]
Looks good, only potential issues are in `prom_metric_family_clone()`, rest of the comments are just for some trivial cleanups.

I'll test this before merge, and let you know whether it fixed my issue.

File: src/write_prometheus.c
Comment (octo) [reply src/write_prometheus.c]: No, metrics are added separately.

✅ Added error handling for the `strdup()` calls.

File: src/write_prometheus.c
Comment (octo) [reply src/write_prometheus.c]: ✅ done

Dito below

File: src/write_prometheus_test.c
Comment (octo) [reply src/write_prometheus_test.c]: ✅ done

File: src/write_prometheus_test.c
Comment (octo) [reply src/write_prometheus_test.c]: ✅ done

File: src/write_prometheus.c
Comment (octo) [reply src/write_prometheus.c]: ✅ rebased

**Comment by eero-t** [comment]
Testing shows prometheus plugin still behaving strangely.  It's not expiring the old data any more:
```
$ curl --no-progress-meter 127.0.0.1:9999/metrics |grep -e TYPE -e HELP -e target | sort | uniq -c | sort -n
      1 # HELP target_info Target metadata
      1 # TYPE target_info gauge
      1 target_info{job="unknown_service:collectd",dev_file="card1",host_name="test-node",pci_bdf="0000:4d:00.0",pci_dev="0x56c0"} 1
  43234 # HELP hw_gpu_temperature_celsius Temperature sensor value (in Celsius) when queried
  43234 # TYPE hw_gpu_temperature_celsius gauge
```

But replying with all the produced metrics, which are _identical_ except for their timestamp:
```
$ curl --no-progress-meter 127.0.0.1:9999/metrics | tail -24
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="global-max"} 46 1708360289834
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="gpu-max"} 46 1708360289834

# HELP hw_gpu_temperature_celsius Temperature sensor value (in Celsius) when queried
# TYPE hw_gpu_temperature_celsius gauge
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="global-max"} 46 1708360301834
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="gpu-max"} 46 1708360301834

# HELP hw_gpu_temperature_celsius Temperature sensor value (in Celsius) when queried
# TYPE hw_gpu_temperature_celsius gauge
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="global-max"} 46 1708360285834
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="gpu-max"} 46 1708360285834

# HELP hw_gpu_temperature_celsius Temperature sensor value (in Celsius) when queried
# TYPE hw_gpu_temperature_celsius gauge
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="global-max"} 46 1708360289834
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="gpu-max"} 46 1708360289834

# HELP hw_gpu_temperature_celsius Temperature sensor value (in Celsius) when queried
# TYPE hw_gpu_temperature_celsius gauge
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="global-max"} 46 1708360283834
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="gpu-max"} 46 1708360283834

# collectd/write_prometheus 6.0.0.git at test-node
```

Notes:
* This amount of metrics is a bit hard to debug because Prometheus plugin outputs metrics in random order, not ordered by timestamp
* Metrics with same timestamp, but from other devices, are still ignored as there's no differentiating "instance" label for them

=> I'll try next whether setting device ID to `service.instance.id` fixes latter issue (as expected).

**Comment by eero-t** [comment]
Ahem, above may have been from older version (there was container with older version running on same node, which may have been answering my queries).

Current one crashes:
```
==1== Invalid read of size 8
==1==    at 0x124EF9: label_pair_compare (metric.c:777)
==1==    by 0x124EF9: label_set_compare (metric.c:791)
==1==    by 0x498BB19: msort_with_tmp.part.0 (msort.c:123)
==1==    by 0x498B882: msort_with_tmp (msort.c:44)
==1==    by 0x498B882: msort_with_tmp.part.0 (msort.c:52)
==1==    by 0x498BE69: msort_with_tmp (msort.c:44)
==1==    by 0x498BE69: qsort_r (msort.c:253)
==1==    by 0x4F58220: prom_metric_family_metric_append (write_prometheus.c:587)
==1==    by 0x4F58220: prom_write (write_prometheus.c:923)
==1==    by 0x118155: plugin_write_thread (plugin.c:870)
==1==    by 0x49DCB59: start_thread (pthread_create.c:444)
==1==    by 0x4A6D463: clone (clone.S:100)
==1==  Address 0x99f9850 is 0 bytes after a block of size 16 alloc'd
==1==    at 0x484A993: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x123F77: label_set_clone (metric.c:237)
==1==    by 0x4F581F7: prom_metric_family_metric_append (write_prometheus.c:578)
==1==    by 0x4F581F7: prom_write (write_prometheus.c:923)
==1==    by 0x118155: plugin_write_thread (plugin.c:870)
==1==    by 0x49DCB59: start_thread (pthread_create.c:444)
==1==    by 0x4A6D463: clone (clone.S:100)
...
==1== Invalid read of size 1
==1==    at 0x484CFE4: strcmp (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0x124F05: label_pair_compare (metric.c:777)
==1==    by 0x124F05: label_set_compare (metric.c:791)
==1==    by 0x498BB19: msort_with_tmp.part.0 (msort.c:123)
==1==    by 0x498B882: msort_with_tmp (msort.c:44)
==1==    by 0x498B882: msort_with_tmp.part.0 (msort.c:52)
==1==    by 0x498BE69: msort_with_tmp (msort.c:44)
==1==    by 0x498BE69: qsort_r (msort.c:253)
==1==    by 0x4F58220: prom_metric_family_metric_append (write_prometheus.c:587)
==1==    by 0x4F58220: prom_write (write_prometheus.c:923)
==1==    by 0x118155: plugin_write_thread (plugin.c:870)
==1==    by 0x49DCB59: start_thread (pthread_create.c:444)
==1==    by 0x4A6D463: clone (clone.S:100)
==1==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==1== 
==1== Process terminating with default action of signal 11 (SIGSEGV): dumping core
```

**Comment by octo** [comment]
> Current one crashes:

Please check if be46fa83410869803554602992dc7251237702d6 fixes the issue.

**Comment by eero-t** [comment]
> > Current one crashes:
> 
> Please check if [be46fa8](https://github.com/collectd/collectd/commit/be46fa83410869803554602992dc7251237702d6) fixes the issue.

Good catch!   I completely missed that.

It fixes the crash, and otherwise plugin seems to work now OK both under Valgrind and without it.  It reports all `target_info`s, and only the last metric states:
```
$ curl --no-progress-meter 127.0.0.1:9999/metrics 
# HELP target_info Target metadata
# TYPE target_info gauge
target_info{job="unknown_service:collectd",instance="0000:4d:00.0",dev_file="card1",host_name="icx-cp-1",pci_bdf="0000:4d:00.0",pci_dev="0x56c0"} 1
target_info{job="unknown_service:collectd",instance="0000:93:00.0",dev_file="card2",host_name="icx-cp-1",pci_bdf="0000:93:00.0",pci_dev="0x56c0"} 1

# HELP hw_gpu_temperature_celsius Temperature sensor value (in Celsius) when queried
# TYPE hw_gpu_temperature_celsius gauge
hw_gpu_temperature_celsius{job="unknown_service:collectd",instance="0000:4d:00.0",location="global-max"} 46 1708451254611
hw_gpu_temperature_celsius{job="unknown_service:collectd",instance="0000:4d:00.0",location="gpu-max"} 46 1708451254611
hw_gpu_temperature_celsius{job="unknown_service:collectd",instance="0000:93:00.0",location="global-max"} 46 1708451254613
hw_gpu_temperature_celsius{job="unknown_service:collectd",instance="0000:93:00.0",location="gpu-max"} 46 1708451254613
```

I'll check that it does something reasonable also when `instance` attribute is missing (and if it does, this is good to go).

**Comment by eero-t** [comment]
And when `instance` attribute is missing, plugin reports:
```
$ curl --no-progress-meter 127.0.0.1:9999/metrics 
# HELP target_info Target metadata
# TYPE target_info gauge
target_info{job="unknown_service:collectd",dev_file="card1",host_name="icx-cp-1",pci_bdf="0000:4d:00.0",pci_dev="0x56c0"} 1
target_info{job="unknown_service:collectd",dev_file="card2",host_name="icx-cp-1",pci_bdf="0000:93:00.0",pci_dev="0x56c0"} 1

# HELP hw_gpu_temperature_celsius Temperature sensor value (in Celsius) when queried
# TYPE hw_gpu_temperature_celsius gauge
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="global-max"} 46 1708452033417
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="gpu-max"} 46 1708452033417
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="global-max"} 46 1708452033417
hw_gpu_temperature_celsius{job="unknown_service:collectd",location="gpu-max"} 46 1708452033417
```

Which is not quite according to: https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/

Because it states that:
* The `service.instance.id` attribute, if present, MUST be converted to the `instance` label; otherwise, `instance` should be added with an empty value
* There MUST be at most one `target_info` metric exported for each unique combination of `job` and `instance`

However, those empty `instance` labels are missing, and there are 2 `target_info` metrics, although there's only single unique `job`+`instance` combo.

I guess those need one more test case...

File: src/write_prometheus.c
Comment (eero-t) [inline src/write_prometheus.c:445]: Maybe also:
```
#ifdef EXPOSE_OPEN_METRICS
    strbuf_print(buf, "target{");
#else
    strbuf_print(buf, "target_info{");
#endif
```
(Or better yet, get the metric name to print from a variable added to earlier ifdef check.)

**Comment by eero-t** [comment]
```
./build-aux/test-driver: line 109: 86323 Segmentation fault      (core dumped) "$@" > $log_file 2>&1
FAIL: test_plugin_write_prometheus
```

**Review by eero-t — APPROVED** [review]
> Is test supposed to **segfault** when it fails? If not, adding hostname to tests fixing segfault, seems like plugin missing some check...

Ah, it's the `strdup(hostname_g)` that segaults, hostname is there by default only with `collectd` daemon, not while testing.

=> Approved

File: src/write_prometheus.c
Comment (octo) [reply src/write_prometheus.c:445]: According to https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems `target_info` is correct. I'm not sure the doc is correct, but I'll hold off on changing this for now.

Comment (eero-t) [reply src/write_prometheus.c:445]: I see, whereas prometheus metric is `target_info` of type `gauge`, OpenMetrics one is `target` metric of type `info`, hence the same name.

**Pull request merged**
